# Day 10 Assignment – Flight Operations Data Analysis

**Organization:** India Data Research Academy (IDRA)

**Topic:** Flight Operations Data Analysis using Python and Pandas

## Objective

The objective of this assignment is to analyze a Flight Operations dataset using Python and Pandas.

The analysis includes understanding the dataset, selecting and filtering data, sorting records, grouping data, performing aggregations, and applying data transformations.

The analysis is used to identify important patterns and findings related to airlines, routes, passengers, delays, weather, booking channels, and passenger satisfaction.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("Day10_Flight_Operations_Dataset.csv")

df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5


In [3]:
df.head(10)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5
5,FL0006,2026-05-20,Akasa Air,Kochi,Delhi,Airbus A319,Economy,123,220,5636,35.0,Delayed,Cloudy,Travel Agency,13,Vegetarian,2
6,FL0007,2026-06-17,Air India Express,Delhi,Srinagar,Airbus A319,Premium Economy,178,210,9430,0.0,On Time,Storm,Online Travel Portal,18,Vegetarian,4
7,FL0008,2026-04-28,Air India,Delhi,Chennai,Airbus A319,Business,157,220,14173,0.0,On Time,Fog,Airline Website,25,Vegan,5
8,FL0009,2026-01-25,Akasa Air,Delhi,Chennai,Airbus A320,Economy,82,210,5893,75.0,Delayed,Cloudy,Mobile App,8,Vegetarian,4
9,FL0010,2026-06-12,Akasa Air,Ahmedabad,Bengaluru,Airbus A321,Economy,125,220,2500,12.0,On Time,Cloudy,Online Travel Portal,23,Non-Vegetarian,4


In [4]:
df.tail()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
175,FL0176,2026-04-10,IndiGo,Ahmedabad,Bengaluru,Airbus A320,Economy,48,189,4298,0.0,On Time,Storm,Mobile App,11,Vegetarian,5
176,FL0177,2026-02-17,Vistara,Chennai,Mumbai,Airbus A320,Business,124,180,17306,20.0,Delayed,Cloudy,Mobile App,8,No Meal,5
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3
178,FL0179,2026-05-08,Vistara,Bengaluru,Kochi,Airbus A320,Economy,134,189,5413,20.0,Delayed,Rain,Online Travel Portal,9,No Meal,2
179,FL0180,2026-01-22,Akasa Air,Delhi,Mumbai,Airbus A321,Economy,144,210,3767,75.0,Delayed,Fog,Online Travel Portal,17,No Meal,5


In [5]:
df.shape

(180, 17)

In [6]:
df.columns

Index(['Flight_ID', 'Flight_Date', 'Airline', 'Origin', 'Destination',
       'Aircraft', 'Travel_Class', 'Passengers', 'Seat_Capacity',
       'Average_Ticket_Price', 'Delay_Minutes', 'Flight_Status', 'Weather',
       'Booking_Channel', 'Avg_Baggage_Kg', 'Meal_Preference',
       'Passenger_Satisfaction'],
      dtype='object')

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Flight_ID               180 non-null    object 
 1   Flight_Date             180 non-null    object 
 2   Airline                 180 non-null    object 
 3   Origin                  180 non-null    object 
 4   Destination             180 non-null    object 
 5   Aircraft                180 non-null    object 
 6   Travel_Class            180 non-null    object 
 7   Passengers              180 non-null    int64  
 8   Seat_Capacity           180 non-null    int64  
 9   Average_Ticket_Price    180 non-null    int64  
 10  Delay_Minutes           173 non-null    float64
 11  Flight_Status           180 non-null    object 
 12  Weather                 180 non-null    object 
 13  Booking_Channel         180 non-null    object 
 14  Avg_Baggage_Kg          180 non-null    in

In [8]:
df.describe()

,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Avg_Baggage_Kg,Passenger_Satisfaction
count,180.000000,180.000000,180.000000,173.000000,180.000000,180.000000
mean,122.772222,197.238889,6111.677778,19.335260,15.150000,3.494444
std,45.635651,15.408176,3615.220040,29.029712,5.705506,1.169908
min,45.000000,180.000000,1949.000000,0.000000,5.000000,2.000000
25%,78.750000,186.000000,3815.500000,0.000000,10.000000,2.000000
50%,125.000000,189.000000,5380.000000,5.000000,15.500000,3.000000
75%,162.250000,210.000000,6790.000000,20.000000,20.000000,5.000000
max,210.000000,220.000000,27832.000000,110.000000,25.000000,5.000000


In [9]:
df.isnull().sum()

,0
Flight_ID,0
Flight_Date,0
Airline,0
Origin,0
Destination,0
Aircraft,0
Travel_Class,0
Passengers,0
Seat_Capacity,0
Average_Ticket_Price,0


In [10]:
df.duplicated().sum()

np.int64(0)

In [11]:
df[["Flight_ID", "Airline", "Origin", "Destination", "Passengers"]].head()

,Flight_ID,Airline,Origin,Destination,Passengers
0,FL0001,SpiceJet,Mumbai,Bengaluru,210
1,FL0002,Akasa Air,Delhi,Mumbai,70
2,FL0003,Air India,Kochi,Delhi,67
3,FL0004,Akasa Air,Srinagar,Delhi,153
4,FL0005,Air India,Pune,Delhi,178


In [12]:
df[["Flight_ID", "Travel_Class", "Average_Ticket_Price"]].head(10)

,Flight_ID,Travel_Class,Average_Ticket_Price
0,FL0001,Economy,6894
1,FL0002,Economy,4468
2,FL0003,Economy,4713
3,FL0004,Economy,3791
4,FL0005,Premium Economy,6505
5,FL0006,Economy,5636
6,FL0007,Premium Economy,9430
7,FL0008,Business,14173
8,FL0009,Economy,5893
9,FL0010,Economy,2500


In [13]:
df[df["Passengers"] > 150]

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5
6,FL0007,2026-06-17,Air India Express,Delhi,Srinagar,Airbus A319,Premium Economy,178,210,9430,0.0,On Time,Storm,Online Travel Portal,18,Vegetarian,4
7,FL0008,2026-04-28,Air India,Delhi,Chennai,Airbus A319,Business,157,220,14173,0.0,On Time,Fog,Airline Website,25,Vegan,5
10,FL0011,2026-04-17,Air India,Mumbai,Bengaluru,Airbus A321,Economy,195,189,3393,5.0,On Time,Clear,Airline Website,17,Vegetarian,2
12,FL0013,2026-05-14,SpiceJet,Pune,Delhi,Airbus A319,Economy,162,220,4689,0.0,On Time,Cloudy,Mobile App,6,Vegetarian,5
20,FL0021,2026-06-23,Vistara,Srinagar,Delhi,Airbus A320,Premium Economy,173,189,9018,0.0,On Time,Fog,Airline Website,14,Vegan,4
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4
28,FL0029,2026-02-06,Air India Express,Mumbai,Bengaluru,Airbus A321,Premium Economy,185,189,8098,0.0,On Time,Rain,Travel Agency,22,Vegan,4


In [14]:
df[df["Flight_Status"] == "Delayed"]

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5
5,FL0006,2026-05-20,Akasa Air,Kochi,Delhi,Airbus A319,Economy,123,220,5636,35.0,Delayed,Cloudy,Travel Agency,13,Vegetarian,2
8,FL0009,2026-01-25,Akasa Air,Delhi,Chennai,Airbus A320,Economy,82,210,5893,75.0,Delayed,Cloudy,Mobile App,8,Vegetarian,4
16,FL0017,2026-05-24,SpiceJet,Delhi,Chennai,Airbus A321,Economy,73,220,4256,20.0,Delayed,Rain,Travel Agency,22,Non-Vegetarian,2
17,FL0018,2026-03-09,Air India Express,Jaipur,Mumbai,Airbus A321,Economy,88,180,6133,20.0,Delayed,Clear,Travel Agency,13,Vegetarian,4
21,FL0022,2026-02-01,Vistara,Kochi,Delhi,Airbus A321,Economy,126,220,5887,75.0,Delayed,Clear,Airline Website,10,Vegetarian,2
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4
29,FL0030,2026-04-10,SpiceJet,Delhi,Srinagar,Airbus A321,Economy,200,210,2716,75.0,Delayed,Storm,Airline Website,10,Vegan,3


In [15]:
df[df["Origin"] == "Delhi"]

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
6,FL0007,2026-06-17,Air India Express,Delhi,Srinagar,Airbus A319,Premium Economy,178,210,9430,0.0,On Time,Storm,Online Travel Portal,18,Vegetarian,4
7,FL0008,2026-04-28,Air India,Delhi,Chennai,Airbus A319,Business,157,220,14173,0.0,On Time,Fog,Airline Website,25,Vegan,5
8,FL0009,2026-01-25,Akasa Air,Delhi,Chennai,Airbus A320,Economy,82,210,5893,75.0,Delayed,Cloudy,Mobile App,8,Vegetarian,4
16,FL0017,2026-05-24,SpiceJet,Delhi,Chennai,Airbus A321,Economy,73,220,4256,20.0,Delayed,Rain,Travel Agency,22,Non-Vegetarian,2
22,FL0023,2026-02-02,Akasa Air,Delhi,Mumbai,Airbus A319,Economy,52,210,4620,12.0,On Time,Fog,Mobile App,9,Non-Vegetarian,2
29,FL0030,2026-04-10,SpiceJet,Delhi,Srinagar,Airbus A321,Economy,200,210,2716,75.0,Delayed,Storm,Airline Website,10,Vegan,3
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4
33,FL0034,2026-06-18,Air India Express,Delhi,Chennai,Boeing 737,Premium Economy,125,186,9278,0.0,On Time,Rain,Online Travel Portal,23,No Meal,3
34,FL0035,2026-04-24,Vistara,Delhi,Srinagar,Boeing 737,Premium Economy,77,210,11441,0.0,On Time,Fog,Travel Agency,9,Vegan,3


In [16]:
df[df["Delay_Minutes"] > 30]

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
5,FL0006,2026-05-20,Akasa Air,Kochi,Delhi,Airbus A319,Economy,123,220,5636,35.0,Delayed,Cloudy,Travel Agency,13,Vegetarian,2
8,FL0009,2026-01-25,Akasa Air,Delhi,Chennai,Airbus A320,Economy,82,210,5893,75.0,Delayed,Cloudy,Mobile App,8,Vegetarian,4
21,FL0022,2026-02-01,Vistara,Kochi,Delhi,Airbus A321,Economy,126,220,5887,75.0,Delayed,Clear,Airline Website,10,Vegetarian,2
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4
29,FL0030,2026-04-10,SpiceJet,Delhi,Srinagar,Airbus A321,Economy,200,210,2716,75.0,Delayed,Storm,Airline Website,10,Vegan,3
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4
32,FL0033,2026-04-27,Akasa Air,Ahmedabad,Bengaluru,Boeing 737,Business,66,180,17625,35.0,Delayed,Cloudy,Mobile App,15,No Meal,4
42,FL0043,2026-06-19,Vistara,Delhi,Chennai,Airbus A319,Economy,64,220,4029,50.0,Delayed,Clear,Mobile App,5,No Meal,3


In [17]:
df[(df["Flight_Status"] == "Delayed") & (df["Passengers"] > 150)]

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4
29,FL0030,2026-04-10,SpiceJet,Delhi,Srinagar,Airbus A321,Economy,200,210,2716,75.0,Delayed,Storm,Airline Website,10,Vegan,3
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4
43,FL0044,2026-04-20,Air India Express,Bengaluru,Hyderabad,Airbus A319,Economy,176,186,3075,35.0,Delayed,Storm,Travel Agency,16,No Meal,3
57,FL0058,2026-04-14,IndiGo,Hyderabad,Kolkata,Boeing 737,Premium Economy,151,210,6834,110.0,Delayed,Cloudy,Travel Agency,16,Vegetarian,4
58,FL0059,2026-01-10,Air India Express,Delhi,Srinagar,Airbus A319,Economy,159,220,3708,20.0,Delayed,Storm,Airline Website,11,No Meal,2
69,FL0070,2026-06-12,SpiceJet,Srinagar,Delhi,Boeing 737,Economy,201,189,2959,20.0,Delayed,Rain,Online Travel Portal,12,Vegan,4


In [18]:
df.sort_values("Passengers", ascending=False).head(10)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4
92,FL0093,2026-04-30,Air India Express,Srinagar,Delhi,Airbus A319,Business,209,220,12970,0.0,On Time,Clear,Travel Agency,8,Vegan,2
148,FL0149,2026-02-10,Akasa Air,Jaipur,Mumbai,Airbus A319,Economy,202,210,6239,5.0,On Time,Rain,Online Travel Portal,8,No Meal,5
69,FL0070,2026-06-12,SpiceJet,Srinagar,Delhi,Boeing 737,Economy,201,189,2959,20.0,Delayed,Rain,Online Travel Portal,12,Vegan,4
153,FL0154,2026-04-09,Air India,Mumbai,Bengaluru,Airbus A321,Economy,201,186,3823,20.0,Delayed,Clear,Online Travel Portal,13,Vegan,2
29,FL0030,2026-04-10,SpiceJet,Delhi,Srinagar,Airbus A321,Economy,200,210,2716,75.0,Delayed,Storm,Airline Website,10,Vegan,3
94,FL0095,2026-06-14,Vistara,Bengaluru,Kochi,Airbus A319,Economy,199,189,4369,0.0,On Time,Fog,Online Travel Portal,11,Vegetarian,5
80,FL0081,2026-04-03,SpiceJet,Delhi,Mumbai,Airbus A320,Economy,199,180,2378,50.0,Delayed,Rain,Online Travel Portal,13,No Meal,4
107,FL0108,2026-05-30,IndiGo,Lucknow,Delhi,Airbus A319,Premium Economy,198,180,10551,75.0,Delayed,Rain,Travel Agency,13,Vegan,2


In [19]:
df.sort_values("Average_Ticket_Price", ascending=False).head(10)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
138,FL0139,2026-01-27,Air India Express,Delhi,Chennai,Airbus A321,Business,169,189,27832,20.0,Delayed,Clear,Airline Website,10,Vegan,4
52,FL0053,2026-05-16,Akasa Air,Mumbai,Bengaluru,Airbus A319,Business,72,210,21499,35.0,Delayed,Clear,Mobile App,15,Non-Vegetarian,2
32,FL0033,2026-04-27,Akasa Air,Ahmedabad,Bengaluru,Boeing 737,Business,66,180,17625,35.0,Delayed,Cloudy,Mobile App,15,No Meal,4
176,FL0177,2026-02-17,Vistara,Chennai,Mumbai,Airbus A320,Business,124,180,17306,20.0,Delayed,Cloudy,Mobile App,8,No Meal,5
108,FL0109,2026-04-12,SpiceJet,Srinagar,Delhi,Airbus A321,Business,129,180,16390,0.0,On Time,Rain,Online Travel Portal,11,Vegan,2
132,FL0133,2026-02-10,IndiGo,Pune,Delhi,Airbus A321,Business,164,180,15324,0.0,On Time,Storm,Travel Agency,20,Vegetarian,4
155,FL0156,2026-03-20,Vistara,Mumbai,Bengaluru,Boeing 737,Business,80,186,15040,0.0,On Time,Storm,Online Travel Portal,20,Vegetarian,3
7,FL0008,2026-04-28,Air India,Delhi,Chennai,Airbus A319,Business,157,220,14173,0.0,On Time,Fog,Airline Website,25,Vegan,5
51,FL0052,2026-03-07,SpiceJet,Jaipur,Mumbai,Airbus A320,Business,163,210,13824,0.0,On Time,Rain,Travel Agency,6,Vegetarian,2
92,FL0093,2026-04-30,Air India Express,Srinagar,Delhi,Airbus A319,Business,209,220,12970,0.0,On Time,Clear,Travel Agency,8,Vegan,2


In [20]:
df.sort_values("Delay_Minutes", ascending=False).head(10)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3
82,FL0083,2026-04-18,Air India Express,Lucknow,Delhi,Airbus A319,Economy,71,180,2484,110.0,Delayed,Cloudy,Online Travel Portal,22,Vegetarian,4
57,FL0058,2026-04-14,IndiGo,Hyderabad,Kolkata,Boeing 737,Premium Economy,151,210,6834,110.0,Delayed,Cloudy,Travel Agency,16,Vegetarian,4
50,FL0051,2026-05-03,Air India,Delhi,Mumbai,Airbus A321,Economy,135,210,5129,110.0,Delayed,Rain,Online Travel Portal,17,Vegetarian,4
128,FL0129,2026-01-04,SpiceJet,Lucknow,Delhi,Boeing 737,Premium Economy,67,210,5657,110.0,Delayed,Cloudy,Online Travel Portal,21,Non-Vegetarian,2
146,FL0147,2026-02-18,IndiGo,Mumbai,Kolkata,Airbus A319,Premium Economy,115,210,9106,110.0,Delayed,Rain,Online Travel Portal,19,Vegetarian,5
169,FL0170,2026-06-25,IndiGo,Mumbai,Kolkata,Airbus A319,Economy,130,180,3938,110.0,Delayed,Rain,Online Travel Portal,17,No Meal,4
161,FL0162,2026-01-18,Air India,Hyderabad,Kolkata,Airbus A321,Economy,128,189,5263,110.0,Delayed,Rain,Mobile App,6,Non-Vegetarian,3
21,FL0022,2026-02-01,Vistara,Kochi,Delhi,Airbus A321,Economy,126,220,5887,75.0,Delayed,Clear,Airline Website,10,Vegetarian,2


In [21]:
print("Total Passengers:", df["Passengers"].sum())
print("Average Passengers per Flight:", df["Passengers"].mean())
print("Highest Passengers on a Flight:", df["Passengers"].max())
print("Lowest Passengers on a Flight:", df["Passengers"].min())

Total Passengers: 22099
Average Passengers per Flight: 122.77222222222223
Highest Passengers on a Flight: 210
Lowest Passengers on a Flight: 45


In [22]:
df["Flight_Status"].value_counts()

,count
Flight_Status,
On Time,114
Delayed,59
Cancelled,7


In [23]:
airline_flights = df.groupby("Airline")["Flight_ID"].count()

airline_flights.sort_values(ascending=False)

,Flight_ID
Airline,
Vistara,37
IndiGo,32
SpiceJet,32
Akasa Air,31
Air India Express,24
Air India,24


In [24]:
airline_analysis = df.groupby("Airline").agg(
    Number_of_Flights=("Flight_ID", "count"),
    Total_Passengers=("Passengers", "sum"),
    Average_Passengers=("Passengers", "mean"),
    Average_Delay=("Delay_Minutes", "mean"),
    Average_Satisfaction=("Passenger_Satisfaction", "mean")
)

airline_analysis.sort_values("Total_Passengers", ascending=False)

,Number_of_Flights,Total_Passengers,Average_Passengers,Average_Delay,Average_Satisfaction
Airline,,,,,
Vistara,37,4484,121.189189,18.628571,3.378378
IndiGo,32,4158,129.937500,25.031250,3.343750
SpiceJet,32,3929,122.781250,24.137931,3.531250
Akasa Air,31,3427,110.548387,14.400000,3.419355
Air India,24,3074,128.083333,21.260870,3.750000
Air India Express,24,3027,126.125000,11.291667,3.666667


In [25]:
origin_analysis = df.groupby("Origin").agg(
    Number_of_Flights=("Flight_ID", "count"),
    Total_Passengers=("Passengers", "sum")
)

origin_analysis.sort_values("Total_Passengers", ascending=False)

,Number_of_Flights,Total_Passengers
Origin,,
Delhi,43,5448
Mumbai,29,3616
Bengaluru,21,2725
Srinagar,15,1957
Kochi,14,1657
Pune,11,1564
Ahmedabad,12,1411
Lucknow,11,1137
Jaipur,8,962


In [26]:
destination_analysis = df.groupby("Destination").agg(
    Number_of_Flights=("Flight_ID", "count"),
    Total_Passengers=("Passengers", "sum")
)

destination_analysis.sort_values("Total_Passengers", ascending=False)

,Number_of_Flights,Total_Passengers
Destination,,
Delhi,51,6315
Bengaluru,28,3344
Mumbai,30,3309
Kolkata,20,2567
Chennai,17,2011
Srinagar,13,1828
Hyderabad,13,1795
Kochi,8,930


In [27]:
class_analysis = df.groupby("Travel_Class").agg(
    Number_of_Flights=("Flight_ID", "count"),
    Total_Passengers=("Passengers", "sum"),
    Average_Ticket_Price=("Average_Ticket_Price", "mean"),
    Average_Satisfaction=("Passenger_Satisfaction", "mean")
)

class_analysis

,Number_of_Flights,Total_Passengers,Average_Ticket_Price,Average_Satisfaction
Travel_Class,,,,
Business,11,1471,16780.000000,3.272727
Economy,134,16054,4550.268657,3.462687
Premium Economy,35,4574,8736.742857,3.685714


In [28]:
weather_analysis = df.groupby("Weather").agg(
    Number_of_Flights=("Flight_ID", "count"),
    Average_Delay=("Delay_Minutes", "mean"),
    Average_Satisfaction=("Passenger_Satisfaction", "mean")
)

weather_analysis.sort_values("Average_Delay", ascending=False)

,Number_of_Flights,Average_Delay,Average_Satisfaction
Weather,,,
Rain,43,22.219512,3.372093
Storm,30,20.103448,3.166667
Cloudy,38,20.028571,3.842105
Clear,34,18.029412,3.352941
Fog,35,15.794118,3.685714


In [29]:
booking_analysis = df.groupby("Booking_Channel").agg(
    Number_of_Flights=("Flight_ID", "count"),
    Total_Passengers=("Passengers", "sum"),
    Average_Ticket_Price=("Average_Ticket_Price", "mean")
)

booking_analysis.sort_values("Total_Passengers", ascending=False)

,Number_of_Flights,Total_Passengers,Average_Ticket_Price
Booking_Channel,,,
Online Travel Portal,50,6175,5636.280000
Travel Agency,44,5542,6223.886364
Mobile App,45,5226,6637.244444
Airline Website,41,5156,5994.170732


In [30]:
df["Flight_Date"] = pd.to_datetime(df["Flight_Date"])

df["Flight_Date"].head()

,Flight_Date
0,2026-03-23
1,2026-06-07
2,2026-05-12
3,2026-05-30
4,2026-02-27


In [31]:
df["Flight_Month"] = df["Flight_Date"].dt.month

df[["Flight_Date", "Flight_Month"]].head()

,Flight_Date,Flight_Month
0,2026-03-23,3
1,2026-06-07,6
2,2026-05-12,5
3,2026-05-30,5
4,2026-02-27,2


In [32]:
df["Flight_Day"] = df["Flight_Date"].dt.day_name()

df[["Flight_Date", "Flight_Day"]].head()

,Flight_Date,Flight_Day
0,2026-03-23,Monday
1,2026-06-07,Sunday
2,2026-05-12,Tuesday
3,2026-05-30,Saturday
4,2026-02-27,Friday


In [33]:
df["Seat_Utilization"] = (
    df["Passengers"] / df["Seat_Capacity"]
) * 100

df[["Flight_ID", "Passengers", "Seat_Capacity", "Seat_Utilization"]].head()

,Flight_ID,Passengers,Seat_Capacity,Seat_Utilization
0,FL0001,210,180,116.666667
1,FL0002,70,210,33.333333
2,FL0003,67,220,30.454545
3,FL0004,153,220,69.545455
4,FL0005,178,180,98.888889


In [34]:
df["Estimated_Revenue"] = (
    df["Passengers"] * df["Average_Ticket_Price"]
)

df[["Flight_ID", "Passengers", "Average_Ticket_Price", "Estimated_Revenue"]].head()

,Flight_ID,Passengers,Average_Ticket_Price,Estimated_Revenue
0,FL0001,210,6894,1447740
1,FL0002,70,4468,312760
2,FL0003,67,4713,315771
3,FL0004,153,3791,580023
4,FL0005,178,6505,1157890


In [35]:
df.sort_values(
    "Estimated_Revenue",
    ascending=False
).head(10)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,...,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Flight_Month,Flight_Day,Seat_Utilization,Estimated_Revenue
138,FL0139,2026-01-27,Air India Express,Delhi,Chennai,Airbus A321,Business,169,189,27832,...,Delayed,Clear,Airline Website,10,Vegan,4,1,Tuesday,89.417989,4703608
92,FL0093,2026-04-30,Air India Express,Srinagar,Delhi,Airbus A319,Business,209,220,12970,...,On Time,Clear,Travel Agency,8,Vegan,2,4,Thursday,95.000000,2710730
132,FL0133,2026-02-10,IndiGo,Pune,Delhi,Airbus A321,Business,164,180,15324,...,On Time,Storm,Travel Agency,20,Vegetarian,4,2,Tuesday,91.111111,2513136
51,FL0052,2026-03-07,SpiceJet,Jaipur,Mumbai,Airbus A320,Business,163,210,13824,...,On Time,Rain,Travel Agency,6,Vegetarian,2,3,Saturday,77.619048,2253312
7,FL0008,2026-04-28,Air India,Delhi,Chennai,Airbus A319,Business,157,220,14173,...,On Time,Fog,Airline Website,25,Vegan,5,4,Tuesday,71.363636,2225161
176,FL0177,2026-02-17,Vistara,Chennai,Mumbai,Airbus A320,Business,124,180,17306,...,Delayed,Cloudy,Mobile App,8,No Meal,5,2,Tuesday,68.888889,2145944
36,FL0037,2026-06-23,Vistara,Bengaluru,Kochi,Airbus A320,Premium Economy,184,220,11614,...,On Time,Fog,Mobile App,5,Vegan,4,6,Tuesday,83.636364,2136976
108,FL0109,2026-04-12,SpiceJet,Srinagar,Delhi,Airbus A321,Business,129,180,16390,...,On Time,Rain,Online Travel Portal,11,Vegan,2,4,Sunday,71.666667,2114310
107,FL0108,2026-05-30,IndiGo,Lucknow,Delhi,Airbus A319,Premium Economy,198,180,10551,...,Delayed,Rain,Travel Agency,13,Vegan,2,5,Saturday,110.000000,2089098
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,...,Delayed,Cloudy,Mobile App,9,Vegetarian,4,1,Thursday,116.111111,2082058


In [36]:
route_analysis = df.groupby(
    ["Origin", "Destination"]
).agg(
    Number_of_Flights=("Flight_ID", "count"),
    Total_Passengers=("Passengers", "sum"),
    Estimated_Revenue=("Estimated_Revenue", "sum")
)

route_analysis.sort_values(
    "Total_Passengers",
    ascending=False
).head(10)

,,Number_of_Flights,Total_Passengers,Estimated_Revenue
Origin,Destination,,,
Delhi,Chennai,17,2011,15801818
Srinagar,Delhi,15,1957,14225022
Mumbai,Bengaluru,16,1933,12021596
Delhi,Srinagar,13,1828,10364358
Bengaluru,Hyderabad,13,1795,8975058
Mumbai,Kolkata,13,1683,9738285
Kochi,Delhi,14,1657,8438384
Delhi,Mumbai,13,1609,7330347
Pune,Delhi,11,1564,10968185


In [37]:
delayed_flights = df[df["Flight_Status"] == "Delayed"]

print("Number of Delayed Flights:", len(delayed_flights))
print("Average Delay:", delayed_flights["Delay_Minutes"].mean())

Number of Delayed Flights: 59
Average Delay: 50.0


In [38]:
df.sort_values(
    "Passenger_Satisfaction",
    ascending=False
).head(10)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,...,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Flight_Month,Flight_Day,Seat_Utilization,Estimated_Revenue
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,...,On Time,Fog,Travel Agency,16,Vegan,5,6,Sunday,33.333333,312760
13,FL0014,2026-06-14,Air India,Lucknow,Delhi,Airbus A321,Economy,78,220,2744,...,On Time,Cloudy,Online Travel Portal,13,No Meal,5,6,Sunday,35.454545,214032
7,FL0008,2026-04-28,Air India,Delhi,Chennai,Airbus A319,Business,157,220,14173,...,On Time,Fog,Airline Website,25,Vegan,5,4,Tuesday,71.363636,2225161
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,...,Delayed,Cloudy,Airline Website,25,No Meal,5,2,Friday,98.888889,1157890
19,FL0020,2026-02-01,SpiceJet,Mumbai,Kolkata,Airbus A320,Economy,125,210,5715,...,On Time,Fog,Travel Agency,24,Non-Vegetarian,5,2,Sunday,59.523810,714375
14,FL0015,2026-01-21,Akasa Air,Lucknow,Delhi,Boeing 737,Economy,133,186,3911,...,On Time,Cloudy,Travel Agency,21,No Meal,5,1,Wednesday,71.505376,520163
12,FL0013,2026-05-14,SpiceJet,Pune,Delhi,Airbus A319,Economy,162,220,4689,...,On Time,Cloudy,Mobile App,6,Vegetarian,5,5,Thursday,73.636364,759618
47,FL0048,2026-01-08,Air India,Delhi,Chennai,Airbus A321,Economy,68,210,3697,...,Delayed,Fog,Mobile App,16,Non-Vegetarian,5,1,Thursday,32.380952,251396
37,FL0038,2026-03-27,Air India,Delhi,Srinagar,Airbus A320,Economy,166,210,5972,...,Cancelled,Rain,Travel Agency,5,Non-Vegetarian,5,3,Friday,79.047619,991352
41,FL0042,2026-03-12,IndiGo,Delhi,Chennai,Boeing 737,Economy,122,186,5670,...,On Time,Storm,Airline Website,21,Vegan,5,3,Thursday,65.591398,691740


## Meaningful Observations

1. The dataset contains 180 flight records with 17 columns.

2. A total of 22,099 passengers are recorded, with an average of approximately 122.77 passengers per flight.

3. Out of 180 flights, 114 flights were On Time, 59 were Delayed, and 7 were Cancelled.

4. Delhi is the busiest origin in the dataset, with 43 flights and 5,448 passengers.

5. Vistara has the highest number of flights in the dataset, with 37 flights.

6. Economy is the most common travel class, with 134 flights and 16,054 passengers.

7. Flights affected by Rain have the highest average delay among the weather conditions, at approximately 22.22 minutes.

8. Travel Agency is the booking channel with the highest total number of passengers, with 5,542 passengers.

## Conclusion

The Flight Operations dataset was successfully analyzed using Python and Pandas.

Different Pandas operations such as data selection, filtering, sorting, groupby(), aggregation, and data transformation were used during the analysis.

The analysis helped identify patterns related to airlines, passengers, routes, travel classes, delays, weather conditions, and booking channels.

The results provide useful insights into flight operations and passenger-related information.

In [39]:
df["Delay_Minutes"] = df["Delay_Minutes"].fillna(
    df["Delay_Minutes"].median()
)

df["Delay_Minutes"].isnull().sum()

np.int64(0)